configuration/configuration.ts

```diff
@@ -0,0 +1,94 @@
+import {FileConfiguration} from "./file-configuration";
+import {EnvConfiguration} from "./env-configuration";
+import {MongoConfiguration} from "./mongo-configuration";
+
+export class Configuration {
+    private static configFilename: string = "user-config.json";
+    private static mongoUriKey: string = "mongoUri";
+    private fileConfig: FileConfiguration;
+    private envConfig: EnvConfiguration;
+    private mongoConfig: MongoConfiguration;
+    private providers: IConfiguration[] = [];
+
+    /**
+     * Asynchronously initialize configuration providers.
+     *
+     * Configuration providers include, in order of preference:
+     *      - JSON file of user preference key-value pairs stored
+     *        in ./user-config.json
+     *      - Environment variables stored with the prefix `azureIot_`
+     *      - Values stored in a MongoDB collection
+     *
+     * Notes:
+     *      - The utilized Mongo collection must contain only a single document.
+     *      - The `mongoUri` key must be given a value before initialization.
+     *
+     * @param {string} collectionName - MongoDB collection to use for
+     * key-value storage.
+     * @param {string[]} requiredKeys - Keys required to be present in
+     * the MongoDB collection.
+     * @param {number} secondsToRetry - Time to wait for Mongo database to
+     * come online and contain requiredKeys, throwing an error after the
+     * allotted time.
+     */
+    public async initialize(
+            collectionName: string = "config",
+            requiredKeys: string[] = [],
+            secondsToRetry: number = 5*60): Promise<void> {
+        this.fileConfig = new FileConfiguration();
+        this.envConfig = new EnvConfiguration();
+        this.mongoConfig = new MongoConfiguration();
+
+        await this.fileConfig.initialize(Configuration.configFilename);
+        this.providers = [this.fileConfig, this.envConfig];
+
+        let mongoUri: string = this.getString(Configuration.mongoUriKey);
+        if (mongoUri === undefined) {
+            throw new Error("Configuration service could not find a mongoUri variable");
+        }
+        await this.mongoConfig.initialize(
+                mongoUri, collectionName, requiredKeys, secondsToRetry);
+
+        this.providers = [this.fileConfig, this.envConfig, this.mongoConfig];
+    }
+
+    /**
+     * Get the value associated with the passed key from the ordered
+     * configuration providers.
+     *
+     * Keys stored as environment variables are stored in the environment
+     * with the prefix `azureIot_`.
+     *
+     * @param {string} key - Name of variable to get.
+     * @return {any} Value of variable named by key.
+     */
+    public getAny(key: string): any {
+        for (let provider of this.providers) {
+            let val: any = provider.getAny(key);
+            if (val !== undefined) {
+                return val;
+            }
+        }
+        return undefined;

```

Comment:
don't explicitly return undefined as the named variable can be redefined...if you omit a return statement the  undefined primitive will be returned